lib/nats/io/client.rb

```diff
@@ -380,6 +376,14 @@ class << self; alias_method :request, :old_request; end
       begin
         srv = select_next_server
 
+        # Use the hostname from the server for TLS hostname verification.
+        if client_using_secure_connection? and single_url_connect_used?
+          # Always reuse the original hostname used to connect.
+          @hostname ||= srv[:hostname]
+        else
+          @hostname = srv[:hostname]
+        end
+
```

Comment:
Moved hostname assignment before actual connection as in case of secure WebSocket connection, it must be made using TLS from the beginning (which required to pass `tls_context` to the socket upon initialization), so hostname must be assigned in advance.